runner/src/Quickstrom/PureScript/Eval/WithObservedStates.hs

```diff
@@ -96,10 +95,28 @@ instance MonadEvalQuery WithObservedStates where
     view (field @"observedStates") >>= \case
       [] -> pure (VBool True)
       _ -> do
-        first' <- require (exprSourceSpan p) (Proxy @"VBool") =<< eval p `catchError` \case
-          Undetermined -> pure (VBool True)
-          e -> throwError e
+        first' <-
+          require (exprSourceSpan p) (Proxy @"VBool") =<< eval p `catchError` \case
+            Undetermined -> pure (VBool True)
+            e -> throwError e
         rest' <-
           require (exprSourceSpan p) (Proxy @"VBool")
             =<< local (field @"observedStates" %~ drop 1) (evalAlways ann p)
         pure (VBool (first' && rest'))
+
+  evalUntil ann p q =
+    view (field @"observedStates") >>= \case
+      [] -> pure (VBool False)
+      _ -> do
+        doesQHold <-
+          require (exprSourceSpan p) (Proxy @"VBool") =<< eval q `catchError` \case
+            Undetermined -> pure (VBool False)
```

Comment:
I'm insure about interpreting `Undetermined` as `False` here. Take an expression like: 

```haskell
always (p `until` q)
```

I don't have any real-world example that makes sense, but I think it should be possible to have a formula like that.

When the finite trace "runs out of states" and the `always` evaluates its condition, it would be false instead of throwing `Undetermined`, and so the entire expression would be false. Quickstrom's `always` isn't "for all states", but rather "for all states that we've observed", and that's why it's interpreting `Undetermined` as true.

So I think these `p` and `q` evaluations should not catch `Undetermined`. That means a proposition like:

```
proposition = p `until` q
```

when verifying a behavior where q never becomes true, would fail with the error message:

```
The formula cannot be determined as there are not enough observed states
```

I'm not sure that error message is optimal, but I think(!) this is the correct semantics. I haven't found any good resource/paper to base this off, especially because of the finiteness.

Reply:
The reason I implemented it like that is that I am unclear on the semantics of `Undetermined`. I assumed it means "we can't evaluate `q` in the current state". If we can't, but we can evaluate `p` to be true, then we should just continue looking further. Also, if we can evaluate `q` to be true now, we can just stop (which is why we're also catching `Undetermined` in the case of `q`).

If it means "there are no states to evaluate", wouldn't we hit the `[]` case before this? Or what does `Undetermined` mean?

Reply:
OK, yeah I see what you're saying. I've used `Undetermined` to mean "there are not enough observed states to evaluate the expression", e.g. a formula like `next true` would be `false` with an empty sequence of states.

I found https://dl.acm.org/doi/10.5555/2540128.2540252, page 2:

![image](https://user-images.githubusercontent.com/1464328/92636563-da257600-f2d7-11ea-984e-58eeb4d03f24.png)

Which if I understand it correctly, means that your implementation is indeed correct. Sorry, my mistake!

Last thing, we might want to evaluate `p` only after we've pattern-matched on `q`? If `q` is true, the no reason to spend time on `p`, right?

Reply:
Thanks for looking into this! I was definitely not sure, so I guess this is yet another time where my unfounded intuition carries me to write a correct implementation! :sweat_smile:

Reply:
well that was an interesting lesson in LTL semantics! 😍